src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonPropertyInfo.cs

```diff
@@ -0,0 +1,411 @@
+                    Assert.Null(prop.Set);
+                    prop.IsRequired = false;
+                    Assert.False(prop.IsRequired);
+                }
+            });
+
+            ClassWithRequiredCustomAttributeAndReadOnlyProperty obj = new();
+
+            string json = """{"SomeProperty":"SomePropertyInitialValue"}""";
+            Assert.Equal(json, JsonSerializer.Serialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(obj, options));
+
+            json = """{"SomeProperty":"SomeOtherValue"}""";
+            var deserialized = JsonSerializer.Deserialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(json, options);
+            Assert.NotNull(deserialized);
+            Assert.Equal("SomePropertyInitialValue", deserialized.SomeProperty);
+
+            json = "{}";
+            deserialized = JsonSerializer.Deserialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(json, options);
+            Assert.NotNull(deserialized);
+            Assert.Equal("SomePropertyInitialValue", deserialized.SomeProperty);
+        }
+
+        [ConditionalFact]
+        public void RequiredReadOnlyPropertyCanBeFixedToBeWritableWithResolver()
+        {
+            JsonSerializerOptions options = CreateOptionsWithModifiers(ti =>
+            {
+                if (ti.Type == typeof(ClassWithRequiredCustomAttributeAndReadOnlyProperty))
+                {
+                    JsonPropertyInfo prop = ti.Properties[0];
+                    Assert.Equal(nameof(ClassWithRequiredCustomAttributeAndReadOnlyProperty.SomeProperty), prop.Name);
+                    Assert.True(prop.IsRequired);
+                    Assert.Null(prop.Set);
+                    prop.Set = (obj, value) => ((ClassWithRequiredCustomAttributeAndReadOnlyProperty)obj).SetSomeProperty((string)value);
+                    Assert.NotNull(prop.Set);
+                }
+            });
+
+            ClassWithRequiredCustomAttributeAndReadOnlyProperty obj = new();
+
+            string json = """{"SomeProperty":"SomePropertyInitialValue"}""";
+            Assert.Equal(json, JsonSerializer.Serialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(obj, options));
+
+            json = """{"SomeProperty":"SomeOtherValue"}""";
+            var deserialized = JsonSerializer.Deserialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(json, options);
+            Assert.NotNull(deserialized);
+            Assert.Equal("SomeOtherValue", deserialized.SomeProperty);
+
+            json = "{}";
+            JsonException exception = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithRequiredCustomAttributeAndReadOnlyProperty>(json, options));
+            Assert.Contains("SomeProperty", exception.Message);
+        }
+
+        public class ClassWithRequiredCustomAttributes
+        {
+            [JsonPropertyOrder(0)]
+            public string NonRequired { get; set; }
+
+            [JsonPropertyOrder(1)]
+            [JsonRequired]
```

Comment:
A few more tests that might be worth adding here:

* virtual properties with `[JsonRequired]` on the base or override.
* properties with both `required` keyword and `[JsonRequired]`. While I do expect that `[JsonRequired]` on `required` is a no-op adding a test to validate that these don't conflict in unexpected ways would be nice.
* `[JsonRequired]` on struct and interface types would be nice to have.

Reply:
for the last bullet - for interfaces we currently ignore properties overall (https://github.com/dotnet/runtime/issues/41749) so there is nothing interesting to test there